### PR TITLE
[GLUTEN-3559][CORE] Fix output partitioning of Hash join

### PR DESCRIPTION
**Pull request by ayushi-agarwal**
## What changes were proposed in this pull request?

As part of this ticket https://issues.apache.org/jira/browse/SPARK-31869 an improvement was added which can decrease the number of exchanges. As output partitioning is overridden in gluten, the test was failing as extra exchange was coming.

(Fixes: \#3559)

## How was this patch tested?

Ran UT locally.

### CODE REVIEW

**Comment by ayushi-agarwal** [comment]
@JkSelf @rui-mo Please review. Thank you

**Comment by ayushi-agarwal** [comment]
@JkSelf @zhli1142015 The build is green. Kindly review.

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala [hunk 1]
```diff
@@ -170,7 +172,7 @@ trait HashJoinLikeExecTransformer
   override def outputPartitioning: Partitioning = joinBuildSide match {
     case BuildLeft =>
       joinType match {
-        case _: InnerLike | RightOuter => right.outputPartitioning
+        case _: InnerLike | RightOuter => expandPartitioning(right.outputPartitioning)
```
Comment (rui-mo) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:175]: HashJoinLikeExecTransformer is also extended by ShuffledHashJoinExecTransformer. Do we need this fix for SHJ?

Comment (ayushi-agarwal) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:175]: Yes, I was looking into that, this change will help it there also so I think we can keep it for SHJ also.

**Comment by ulysses-you** [comment]
it seems the code is difference since Spark3.4, shall we move this to shim module ?

File: gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala [hunk 2]
```diff
@@ -179,14 +181,84 @@ trait HashJoinLikeExecTransformer
...
+          expandPartitioning(left.outputPartitioning)
         case RightOuter => right.outputPartitioning
         case x =>
           throw new IllegalArgumentException(
             s"HashJoin should not take $x as the JoinType with building right side")
       }
   }
 
+  // https://issues.apache.org/jira/browse/SPARK-31869
+  // ToDo: https://issues.apache.org/jira/browse/SPARK-45882
```
Comment (ayushi-agarwal) [inline gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:193]: I have added a ToDO here, for this https://issues.apache.org/jira/browse/SPARK-45882 we would need to add in shim. Shall that be taken in separate PR?

Comment (ayushi-agarwal) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:193]: @ulysses-you Shall we move this to shim module as part of this in separate PR?

Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:193]: I'd like to solve this todo in this pr, as we need to move the whole code to shim module.

Comment (ayushi-agarwal) [reply gluten-core/src/main/scala/io/glutenproject/execution/HashJoinExecTransformer.scala:193]: Thanks @ulysses-you I will update the PR

**Review by ulysses-you — APPROVED** [review]
thank you @ayushi-agarwal , lgtm. cc @rui-mo if you have other comments

**Comment by zzcclp** [comment]
LGTM

**Review by zhli1142015 — APPROVED** [review]

**Comment by ayushi-agarwal** [comment]
Thank you @zzcclp @ulysses-you for reviewing.
